src/core/fetchers/segment/promise_prioritizer.ts

```diff
@@ -0,0 +1,321 @@
+      priority < this._minPendingPriority
+    ) {
+      this._minPendingPriority = priority;
+    } else if (this._minPendingPriority === prevPriority) {
+      // was highest priority
+      if (this._pendingTasks.length === 1) {
+        this._minPendingPriority = priority;
+      } else {
+        this._minPendingPriority = Math.min(
+          ...this._pendingTasks.map((t) => t.priority)
+        );
+      }
+      this._loopThroughWaitingQueue();
+    } else {
+      // We updated a task which already had a priority value higher than the
+      // minimum to a value still superior to the minimum. Nothing can happen.
+      return;
+    }
+
+    if (this._isRunningHighPriorityTasks()) {
+      // Always interrupt cancellable tasks after all other side-effects, to
+      // avoid re-entrancy issues
+      this._interruptCancellableTasks();
+    }
+  }
+
+  private _loopThroughWaitingQueue(): void {
+    const minWaitingPriority = this._waitingQueue.reduce(
+      (acc: number | null, elt) => {
+        return acc === null || acc > elt.priority ? elt.priority : acc;
+      },
+      null
+    );
+    if (
+      minWaitingPriority === null ||
+      (this._minPendingPriority !== null &&
+        this._minPendingPriority < minWaitingPriority)
+    ) {
+      return;
+    }
+    for (let i = 0; i < this._waitingQueue.length; i++) {
+      const priorityToCheck =
+        this._minPendingPriority === null
+          ? minWaitingPriority
+          : Math.min(this._minPendingPriority, minWaitingPriority);
+      const elt = this._waitingQueue[i];
+      if (elt.priority <= priorityToCheck) {
+        this._startWaitingQueueTask(i);
+        i--; // previous operation should have removed that element from the
+        // the waiting queue
+      }
+    }
+  }
+
+  private _interruptCancellableTasks(): void {
+    this._pendingTasks
+      .filter((pendingTask) => pendingTask.priority >= this._prioritySteps.low)
+      .forEach((task) => {
+        this._interruptPendingTask(task);
+        return this._interruptCancellableTasks();
+      });
+  }
+
+  private _startWaitingQueueTask(index: number): void {
+    const task = this._waitingQueue.splice(index, 1)[0];
+    task.trigger(true);
+  }
+
+  private _interruptPendingTask(task: IPrioritizerTask<T>): void {
+    const pendingTasksIndex = arrayFindIndex(
+      this._pendingTasks,
+      (elt) => elt.promise === task.promise
+    );
+    if (pendingTasksIndex < 0) {
+      log.warn("FP: Interrupting a non-existent pending task. Aborting...");
+      return;
+    }
+
+    // Stop task and put it back in the waiting queue
+    this._pendingTasks.splice(pendingTasksIndex, 1);
+    this._waitingQueue.push(task);
+    if (this._pendingTasks.length === 0) {
+      this._minPendingPriority = null;
+    } else if (this._minPendingPriority === task.priority) {
+      this._minPendingPriority = Math.min(
+        ...this._pendingTasks.map((t) => t.priority)
+      );
+    }
+    task.cancel();
```

Comment:
I don't think this actually cancels the task.

From what I can see it does actually nothing at all.